dropshot/src/router.rs

```diff
@@ -139,6 +141,19 @@ impl PathSegment {
     }
 }
 
+/**
+ * Wrapper for a path that's the result of user input i.e. an HTTP query.
+ * We use this type to avoid confusion with paths used to define routes.
+ */
+#[derive(Debug)]
+pub struct InputPath<'a>(&'a str);
```

Comment:
I have no issue with this current use of typing, but I feel like I've usually seen the strong types used as an "output" of parsing functions, rather than input. IMO that seems more useful, since it propagates the notion that "this `&str` has been validated", whereas I think `InputPath` is currently truly indistinguishable from an arbitrary `&str`.

(I'm kinda eyeing `route_path_to_segments` and `input_path_to_segments` below - it does seem like both could take an arbitrary `&str` as input, but the results kinda have different implications)

Reply:
I don't have a strong feeling about what would be most idiomatic here. I wanted to "quarantine" the input type to avoid accidental use of it. I agree that it's all a little meh. I'd love any input regarding how to improve it.

Reply:
I wanna give the caveat - I think this patch is decidedly a net improvement, so this comment shouldn't be a blocker!

The reason I'm a little skeptical of the current usage is that I don't think the quarantining is actually happening very clearly (aside from requiring a bunch of `.into()` calls). Any function that operates on an `InputPath` *may as well* operate on a `&str`, since there is (seemingly?) no check or consideration to distinguish them.

IMO, the path that's the result of a "user input" actually *is* a str. If we want to distinguish that from a route-defining path, which has a higher burden of validation, perhaps *that* is the type we should abstract more cautiously.

As an example, `route_path_to_segments` returns a `Vec<&str>` - but it implicitly carries the implication that "this is actually a non-empty `&str` type, without `/` characters.

We could make that more explicit by returning a `Vec<Segment>`, where *Segment* is a `&str`, but it can be only constructed during validation.

Similarly, `input_path_to_segments` could return a `Vec<Segment>`, to indicate that it represents a percent-decoded, non "./.." str.

(This actually raises the question - should these functions actually be returning distinct `Segment` types? They appear to be performing slightly different validations)

Reply:
FWIW: I think this pattern is generally the one Cliff calls "the typestate pattern".

- http://cliffle.com/blog/rust-typestate/
- https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/